File: rust_problem_solving/src/recursion/factorial.rs
pub mod caluclate_factorial {

    // time O(n) , space O(1)
    pub fn iterative(n: u32) {
        let mut res = 1;
        for i in 1..=n {
            res = res * i;
        }
        println!("{}", res);
    }

    // time O(n) , space O(n)
    pub fn recursive(n: u32) -> u32 {
        if n == 1 {
            return 1;
        }

        n * recursive(n - 1)
    }
}

File: python_go/maths/factorial.py

# Recursive approach
# Time Complexity: O(n)
# Auxiliary Space: O(n)
def factorialxR(n):
	
	# single line to find factorial
	return 1 if (n==1 or n==0) else n * factorialxR(n - 1);


# Iterative approach
# Time Complexity: O(n)
# Auxiliary Space: O(1)
def factorialxI(n):
	if n < 0:
		return 0
	elif n == 0 or n == 1:
		return 1
	else:
		fact = 1
		while(n > 1):
			fact *= n
			n -= 1
		return fact

